drivers/nec/display/display.cr

```diff
@@ -0,0 +1,397 @@
+      volume_status(20)
+    end
+  end
+
+  def power?(**options)
+    do_send(MsgType::Command, Command::Power_query, **options)
+  end
+
+  def switch_to(input : Input)
+    @target_input = input
+    @target_audio = nil
+
+    data = Command::Video_input.to_s + input.to_s
+
+    do_send(MsgType::Set_parameter, data, name: "input", delay: 6.seconds)
+    video_input
+
+    # Double check the input again!
+    # @input_double_check.cancel if @input_double_check
+    # @input_double_check = schedule.in(4.seconds) do
+    #   @input_double_check = nil
+    #   video_input
+    # end
+    # TODO: check if the below is equivalent to what's above
+    @input_double_check.try &.terminate
+    schedule.in(4.seconds) do
+      @input_double_check = nil
+      video_input.as(PlaceOS::Driver::Transport)
+    end
+
+    logger.debug { "-- NEC LCD, requested to switch to: #{input}" }
+  end
+
+  enum Audio
+    Audio1       = 1
+    Audio2       = 2
+    Audio3       = 3
+    Hdmi         = 4
+    Tv           = 6
+    Display_port = 7
+
+    def to_s : String
+      Nec::Display::All.format_value(self.value)
+    end
+  end
+
+  def switch_audio(input : Audio)
+    @target_audio = input
+
+    data = Command::Audio_input.to_s + input.to_s
+
+    do_send(MsgType::Set_parameter, data, name: "audio")
+    mute_status(20) # higher status than polling commands - lower than input switching
+    volume_status(20)
+
+    logger.debug { "-- NEC LCD, requested to switch audio to: #{input}" }
+  end
+
+  def auto_adjust
+    data = Command::Auto_setup.to_s + "0001"
+    # TODO: find out if there is an equivalent for delay_on_receive
+    do_send(MsgType::Set_parameter, data)#, delay_on_receive: 4.seconds)
+  end
+
+  def brightness(val : Int32)
+    data = Command::Brightness_status.to_s + self.class.format_value(val.clamp(0, 100))
+
+    do_send(MsgType::Set_parameter, data, name: "brightness")
+    do_send(MsgType::Command, "0C", name: "brightness_save") # Save the settings
+  end
+
+  def contrast(val : Int32)
+    data = Command::Contrast_status.to_s + self.class.format_value(val.clamp(0, 100))
+
+    do_send(MsgType::Set_parameter, data, name: "contrast")
+    do_send(MsgType::Command, "0C", name: "contrast_save") # Save the settings
+  end
+
+  def volume(val : Int32)
+    data = Command::Volume_status.to_s + self.class.format_value(val.clamp(0, 100))
+
+    do_send(MsgType::Set_parameter, data, name: "volume_status")
+    do_send(MsgType::Command, "0C", name: "volume_save") # Save the settings
+    self[:audio_mute] = false # audio is unmuted when the volume is set
+  end
+
+  def mute_audio(state : Bool = true, index : Int32 | String = 0)
+    data = Command::Mute_status.to_s + (state ? "0001" : "0000")
+
+    do_send(MsgType::Set_parameter, data)
+    logger.debug { "requested to update mute to #{state}" }
+  end
+
+  def unmute_audio
+    mute_audio(false)
+  end
+
+  # LCD Response code
+  def received(data, task)
+    ascii_string = String.new(data)
```

Comment:
You may find it useful to use [`String#hexbytes`](https://crystal-lang.org/api/0.35.1/String.html#hexbytes:Bytes-instance-method) here. This will allow you to map back from the hex encoded ascii string, to the byte array that it presents.

Reply:
I'm not sure how I can use that. Isn't data of type `Bytes`? 
![image](https://user-images.githubusercontent.com/18463731/93620124-6c610480-f9d1-11ea-948a-877657f20448.png)

Reply:
With this protocol the incoming data will be a `Bytes` containing an ASCII representation of the actual response packet. If you first read this into a string, then dump back out, this should decode into a suitable format for parsing.

```crystal
response = String.new(data).hexbytes
```

There's a more efficient way to do this which drop the need for that intermediate string, but the readability tradeoff is likely not worth the minuscule perf boost.

Reply:
I don't think I can do that @KimBurgess as not all of the data that comes back is actually an ascii string and it's a mix of ascii values and hex values https://play.crystal-lang.org/#/r/9sl0

Reply:
Ah yep, if that's an example of the response there's a few delimiters in there that would need to be used to segment the packet into a few different field first. If the current approach works then stick with that.